### DISCUSSION TITLE
Using `File` to limit `UploadFile`'s size should work but doesn't.

### DISCUSSION THREAD

**Discussion by Diegovsky** [fastapi/fastapi#11750] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated
from fastapi import FastAPI, UploadFile, File

app = FastAPI()

@app.post("/")
async def echo(file: Annotated[UploadFile, File(max_length=1000)]):
    print(await file.read())
```

### Description

According to [this section in the official docs](https://fastapi.tiangolo.com/tutorial/request-files/?h=upload#multiple-file-uploads-with-additional-metadata), `File` is used to tag additional information about `UploadFile`, however, it seems that setting `max_length` makes it not work, resulting in the following (reduced) error:
```
(snip)
File "/tmp/tmp.mWJ9k2fo2B/.direnv/python-3.12/lib/python3.12/site-packages/pydantic/_internal/_validators.py", line 278, in max_length_validator
    if len(x) > max_length:
```

### Operating System

Linux

### FastAPI Version

0.111.0

### Pydantic Version

2.7.4

### Python Version

3.12.4

**Comment by ceb10n** [comment]
When you import File, you are using a function that will return a File class, that inherits from Form. And you get generic validations that only applies to certain types.

For example, you can pass `allow_inf_nan=True` or `decimal_places=2`, but it is only applicable for numbers. 

In your case, you are trying to use `max_length`, that according to the [docs/source code](https://github.com/tiangolo/fastapi/blob/master/fastapi/param_functions.py#L2036), is only applicable to strings:

```python
def File(  # noqa: N802
    default: Annotated[
        Any,
        Doc(
            """
            Default value if the parameter field is not set.
            """
        ),
    ] = Undefined,
    *,
   # other params
    max_length: Annotated[
        Optional[int],
        Doc(
            """
            Maximum length for strings.
            """
        ),
```

**Reply by Diegovsky** [reply]
Thanks, that does explain a lot :)

**Comment by chyok** [comment, marked as answer]
Hi @Diegovsky ,
you can try this:
```python
@app.post("/")
async def echo(file: Annotated[UploadFile, File()],
               content_length: Annotated[int | None, Header(lt=1000*1024)] = None):
```
We can use the Content-Length request header to pre-validate the file size (the unit is Bytes). For example, the code you provided implemented a check to ensure the file size is less than 1000KB. Additionally, you may notice that in the Swagger UI, the browser will automatically calculate and populate this header for you. However, please note that for other clients, you need to ensure they properly send the Content-Length header.

By the way, you can directly use file.size to determine the file size (a property of the UploadFile). However, by the time the server receives the request, the file data has already been transmitted. If you want to save network bandwidth, the above approach that validate the Content-Length header before accepting the file, is more suitable for you.

**Reply by Diegovsky** [reply]
While this does solve my problem, it would be great if the docs mentioned this.

**Reply by chyok** [reply]
I think this is a general business solution, rather than something that should be covered in the FastAPI documentation.

**Reply by Diegovsky** [reply]
You're right about this, but the docs mention using `File` as *the* metadata for `UploadFile`, and this doesn't work like one would assume. I think there should be a section that warns against using `File`'s validators
